src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -509,19 +402,27 @@ InsertRowsResponse insertBufferedRecords(StreamingBuffer streamingBufferToInsert
       this.previousFlushTimeStampMs = System.currentTimeMillis();
 
       LOGGER.info(
-          "Successfully called insertRows for channel:{}, buffer:{}, insertResponseHasErrors:{}",
+          "Successfully called insertRows for channel:{}, buffer:{}, insertResponseHasErrors:{},"
+              + " needToResetOffset:{}",
           this.getChannelName(),
           streamingBufferToInsert,
-          response.hasErrors());
+          response.hasErrors(),
+          response.needToResetOffset());
       if (response.hasErrors()) {
         handleInsertRowsFailures(
             response.getInsertErrors(), streamingBufferToInsert.getSinkRecords());
       }
 
-      // Check whether we need to reset the offset due to schema evolution
+      // Due to schema evolution, we may need to reopen the channel and reset the offset in kafka
+      // since it's possible that not all rows are ingested
       if (response.needToResetOffset()) {
-        streamingApiFallbackSupplier(
-            StreamingApiFallbackInvoker.INSERT_ROWS_SCHEMA_EVOLUTION_FALLBACK);
+        final long offsetRecoveredFromSnowflake =
+            streamingApiFallbackSupplier(
+                StreamingApiFallbackInvoker.INSERT_ROWS_SCHEMA_EVOLUTION_FALLBACK);
+        // If there is no valid offset token at server side even after the reset, retry it again
+        if (offsetRecoveredFromSnowflake == NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE) {
+          insertBufferedRecords(streamingBufferToInsert);
```

Comment:
for my understanding, Is the reason we are retrying is because we have altered the schema in fallback supplier above?